design/design-doc-webhooks.md

```diff
@@ -0,0 +1,383 @@
+
+[Mutation](https://github.com/kubernetes-sigs/controller-runtime/blob/0208f43/pkg/webhook/admission/defaulter.go#L28)
+webhook:
+```golang
+// Defaulter defines functions for setting defaults on resources
+type Defaulter interface {
+  runtime.Object
+  Default()
+}
+```
+
+[Validation](https://github.com/kubernetes-sigs/controller-runtime/blob/8161d1c/pkg/webhook/admission/validator.go#L30)
+webhook:
+```golang
+// Validator defines functions for validating an operation
+type Validator interface {
+  runtime.Object
+  ValidateCreate() error
+  ValidateUpdate(old runtime.Object) error
+  ValidateDelete() error
+}
+```
+
+[Conversion](https://github.com/kubernetes-sigs/controller-runtime/blob/836b363d731dbbea5591db3c7ca2a36c2dd6c151/pkg/conversion/conversion.go#L22)
+webhook:
+```golang
+// Convertible defines capability of a type to convertible i.e. it can be converted to/from a hub type.
+type Convertible interface {
+  runtime.Object
+  ConvertTo(dst Hub) error
+  ConvertFrom(src Hub) error
+}
+
+// Hub marks that a given type is the hub type for conversion. This means that
+// all conversions will first convert to the hub type, then convert from the hub
+// type to the destination type. All types besides the hub type should implement
+// Convertible.
+type Hub interface {
+  runtime.Object
+  Hub()
+}
+```
+
+We see that the interfaces include `runtime.Object`, which means the functions
+need to be members of the kinds, i.e. `RDSInstance` struct needs to implements
+these functions. The abstraction layer that accepts `Defaulter` and `Validator`
+look pretty similar so I'll copy only the [mutation
+one](https://github.com/kubernetes-sigs/controller-runtime/blob/0208f43/pkg/webhook/admission/defaulter.go#L54):
+
+```golang
+// Handle handles admission requests.
+func (h *mutatingHandler) Handle(ctx context.Context, req Request) Response {
+  if h.defaulter == nil {
+    panic("defaulter should never be nil")
+  }
+
+  // Get the object in the request
+  obj := h.defaulter.DeepCopyObject().(Defaulter)
+  err := h.decoder.Decode(req, obj)
+  if err != nil {
+    return Errored(http.StatusBadRequest, err)
+  }
+
+  // Default the object
+  obj.Default()
+  marshalled, err := json.Marshal(obj)
+  if err != nil {
+    return Errored(http.StatusInternalServerError, err)
+  }
+
+  // Create the patch
+  return PatchResponseFromRaw(req.Object.Raw, marshalled)
+}
+```
+
+The conversion webhook implementation is a bit [more
+complex](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/webhook/conversion/conversion.go#L22)
+due to the marshalling and version checks that need to happen. However, it seems
+that some of the cases it needs to handle [are not
+necessary](https://github.com/kubernetes-sigs/controller-runtime/issues/1388) in
+v1 CRD mechanics.
+
+These high level abstractions are fairly simple but they have a few drawbacks:
+* They require implementation in API type packages and we want to keep API type
```

Comment:
I discovered that with https://github.com/kubernetes-sigs/controller-runtime/pull/1676 now it's possible to register custom handlers that are not part of the API definition, so this concern is mostly gone. I'll update this PR once I'm done with prototyping to see how it all looks like.

Reply:
I have decided to change the proposal from `controller-runtime high level abstraction does not work for us so we will build one` to `we will build a lean layer on top of what is in controller-runtime so that we can append multiple function that will be called in a single webhook call` given that the biggest concern about their abstraction layer is eliminated with custom handlers.